File: Group-Video/OpenVideoCall-iOS-Objective-C/OpenVideoCall/SettingsViewController.m
//
//  SettingsViewController.m
//  OpenVideoCall
//
//  Created by GongYuhua on 2016/9/12.
//  Copyright © 2016年 Agora. All rights reserved.
//

#import "SettingsViewController.h"
#import "DimensionCell.h"
#import <AgoraRtcEngineKit/AgoraRtcEngineKit.h>

@interface SettingsViewController () <UICollectionViewDelegate, UICollectionViewDataSource, UIPickerViewDelegate, UIPickerViewDataSource>
@property (weak, nonatomic) IBOutlet UICollectionView *dimensionCollectionView;
@property (weak, nonatomic) IBOutlet UIPickerView *frameRatePickerView;
@property (weak, nonatomic) IBOutlet UILabel *frameRateLabel;

@property (assign, nonatomic) AgoraVideoFrameRate selectedFrameRate;
@property (assign, nonatomic) CGSize selectedDimension;
@property (strong, nonatomic) Settings *settings;
@property (strong, nonatomic) NSArray *dimensionList;
@property (strong, nonatomic) NSArray *frameRateList;
@end

@implementation SettingsViewController
#pragma mark - Getter, Setter
- (NSArray *)dimensionList {
    if (!_dimensionList) {
        _dimensionList = @[@(AgoraVideoDimension160x120),
                           @(AgoraVideoDimension240x180),
                           @(AgoraVideoDimension320x240),
                           @(AgoraVideoDimension640x360),
                           @(AgoraVideoDimension640x480),
                           @(AgoraVideoDimension960x720)];
    }
    return _dimensionList;
}

- (NSArray *)frameRateList {
    if (!_frameRateList) {
        _frameRateList = @[@(AgoraVideoFrameRateFps1),
                           @(AgoraVideoFrameRateFps7),
                           @(AgoraVideoFrameRateFps10),
                           @(AgoraVideoFrameRateFps15),
                           @(AgoraVideoFrameRateFps24),
                           @(AgoraVideoFrameRateFps30)];
    }
    return _frameRateList;
}

- (Settings *)settings {
    return [self.dataSource settingsVCNeedSettings];
}

- (void)setSelectedDimension:(CGSize)selectedDimension {
    _selectedDimension = selectedDimension;
    [self.dimensionCollectionView reloadData];
}

- (void)setSelectedFrameRate:(AgoraVideoFrameRate)selectedFrameRate {
    _selectedFrameRate = selectedFrameRate;
    self.frameRateLabel.text = [self descriptionOfFrameRate:selectedFrameRate];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self updateCollectionViewLayout];
    self.selectedDimension = self.settings.dimension;
    self.selectedFrameRate = self.settings.frameRate;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSInteger frameRateChooseIndexSection = 1;
    if (indexPath.section == frameRateChooseIndexSection) {
        [tableView deselectRowAtIndexPath:indexPath animated:YES];
        [self doFrameRatePressed];
    }
}

#pragma mark - Private
- (void)doFrameRatePressed {
    self.frameRatePickerView.hidden = !self.frameRatePickerView.hidden;
    NSUInteger row = [self.frameRateList indexOfObject:@(self.selectedFrameRate)];
    [self.frameRatePickerView selectRow:row inComponent:0 animated:YES];
}

- (void)updateCollectionViewLayout {
    CGFloat itemInteritemSpacing = 8;
    CGFloat collectionViewInteritemSpacing = 19;
    CGFloat screenWidth = [UIScreen mainScreen].bounds.size.width;
    CGFloat width = ((screenWidth - (collectionViewInteritemSpacing * 2) - (itemInteritemSpacing * 2)) / 3);
    CGFloat height = 40;
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    layout.minimumLineSpacing = 16;
    layout.minimumInteritemSpacing = itemInteritemSpacing;
    layout.itemSize = CGSizeMake(width, height);
    [self.dimensionCollectionView setCollectionViewLayout:layout animated:YES];
}

#pragma mark - UICollectionViewDelegate, UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.dimensionList.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    DimensionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"DimensionCell" forIndexPath:indexPath];
    CGSize dimension = [self.dimensionList[indexPath.row] CGSizeValue];
    BOOL isSelected = [self sizeIsEqual:dimension right:self.selectedDimension];
    [cell updateWithDimension:dimension isSelected:isSelected];
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    self.selectedDimension = [self.dimensionList[indexPath.row] CGSizeValue];
}

#pragma mark - UIPickerViewDelegate, UIPickerViewDataSource
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return self.frameRateList.count;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    self.selectedFrameRate = [self.frameRateList[row] integerValue];
    self.frameRatePickerView.hidden = YES;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return [self descriptionOfFrameRate:[self.frameRateList[row] integerValue]];
}

#pragma mark - Others
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 16;
}

- (NSString *)descriptionOfFrameRate:(AgoraVideoFrameRate)frameRate {
    switch (frameRate) {
        case AgoraVideoFrameRateFps1:  return @"1 fps";     break;
        case AgoraVideoFrameRateFps7:  return @"7 fps";     break;
        case AgoraVideoFrameRateFps10: return @"10 fps";    break;
        case AgoraVideoFrameRateFps15: return @"15 fps";    break;
        case AgoraVideoFrameRateFps24: return @"24 fps";    break;
        case AgoraVideoFrameRateFps30: return @"30 fps";    break;
        default:                       return @"unsported"; break;
    }
}

- (BOOL)sizeIsEqual:(CGSize)left right:(CGSize)right {
    if (left.width == right.width && left.height == right.height) {
        return YES;
    }
    return NO;
}
@end

File: Group-Video/OpenVideoCall-iOS/OpenVideoCall/SettingsViewController.swift
//
//  SettingsViewController.swift
//  OpenVideoCall
//
//  Created by GongYuhua on 6/25/16.
//  Copyright © 2016 Agora. All rights reserved.
//

import UIKit
import AgoraRtcEngineKit

protocol SettingsVCDataSource: NSObjectProtocol {
    func settingsVCNeedSettings() -> Settings
}

protocol SettingsVCDelegate: NSObjectProtocol {
    func settingsVC(_ vc: SettingsViewController, didSelect dimension: CGSize)
    func settingsVC(_ vc: SettingsViewController, didSelect frameRate: AgoraVideoFrameRate)
}

class SettingsViewController: UITableViewController {
    
    @IBOutlet weak var frameRateLabel: UILabel!
    @IBOutlet weak var frameRatePickerView: UIPickerView!
    @IBOutlet weak var dimensionCollectionView: UICollectionView!
    
    private var settings: Settings {
        return dataSource!.settingsVCNeedSettings()
    }
    
    private var selectedDimension: CGSize = CGSize.defaultDimension() {
        didSet {
            dimensionCollectionView?.reloadData()
        }
    }
    
    private var selectedFrameRate: AgoraVideoFrameRate = AgoraVideoFrameRate.defaultValue {
        didSet {
            frameRateLabel?.text = selectedFrameRate.description
        }
    }
    
    private let dimensionList: [CGSize] = CGSize.validDimensionList()
    private let frameRateList: [AgoraVideoFrameRate] = AgoraVideoFrameRate.validList()
    
    weak var delegate: SettingsVCDelegate?
    weak var dataSource: SettingsVCDataSource?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        selectedDimension = settings.dimension
        selectedFrameRate = settings.frameRate
        updateCollectionViewLayout()
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let frameRateChooseIndexSection = 1
        if indexPath.section == frameRateChooseIndexSection {
            tableView.deselectRow(at: indexPath, animated: true)
            doFrameRatePressed()
        }
    }
    
    deinit {
        delegate?.settingsVC(self, didSelect: selectedDimension)
        delegate?.settingsVC(self, didSelect: selectedFrameRate)
    }
}

private extension SettingsViewController {
    func doFrameRatePressed() {
        frameRatePickerView.isHidden.toggle()
        if let row = frameRateList.index(of: selectedFrameRate) {
            frameRatePickerView.selectRow(row, inComponent: 0, animated: true)
        }
    }
    
    func updateCollectionViewLayout() {
        let itemInteritemSpacing: CGFloat = 8
        let collectionViewInteritemSpacing: CGFloat = 19
        let width = Double((UIScreen.main.bounds.width - (collectionViewInteritemSpacing * 2) - (itemInteritemSpacing * 2)) / 3)
        let height = 40.0
        let layout = UICollectionViewFlowLayout()
        layout.minimumLineSpacing = 16
        layout.minimumInteritemSpacing = itemInteritemSpacing
        layout.itemSize = CGSize(width: width, height: height)
        dimensionCollectionView.setCollectionViewLayout(layout, animated: true)
    }
}

extension SettingsViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return dimensionList.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "DimensionCell", for: indexPath) as! DimensionCell
        let dimension = dimensionList[indexPath.row]
        cell.update(with: dimension, isSelected: (dimension == selectedDimension))
        return cell
    }
}

extension SettingsViewController: UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let dimension = dimensionList[indexPath.row]
        selectedDimension = dimension
    }
}

extension SettingsViewController: UIPickerViewDataSource {
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return frameRateList.count
    }
}

extension SettingsViewController: UIPickerViewDelegate {
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        let frameRate = frameRateList[row]
        selectedFrameRate = frameRate
        frameRatePickerView.isHidden.toggle()
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return frameRateList[row].description
    }
}

extension SettingsViewController {
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 16
    }
}
